Entities/AtomGroup.cpp

```diff
@@ -1,2618 +1,1750 @@
+			// Move back before gibbing so gibs don't end up inside terrain
+			intersectedMO->SetPos(intersectedMO->GetPos() - intersectedExit);
+			intersectedMOS->GibThis(totalExitVector);
+		}
+
+		// TODO: this isn't really true since we don't check for clearness after moving the position
+		return intersectingAtoms.empty();
 	}
 
-    // Restart and go through all atoms to find all intersecting the specific intersected MO
-    for (Atom *atom : m_Atoms)
-    {
-        atomPos = atom->GetCurrentPos();
-        if (g_SceneMan.GetMOIDPixel(atomPos.GetFloorIntX(), atomPos.GetFloorIntY()) == currentMOID)
-        {
-            // Add atom to list of intersecting ones
-            intersectingAtoms.push_back(atom);
-        }
-    }
-
-    // No intersections - we're clear?!
-//        if (intersectingAtoms.empty())
-//            return true;
-//        RTEAssert(!intersectingAtoms.empty(), "Couldn't find any intersections after finding one?!");
-
-    // If all atoms are intersecting, we're screwed?!
-//        if (intersectingAtoms.size() >= m_Atoms.size())
-//            return false;
-
-    // Go through all intesecting atoms and find their average inverse normal
-	for (const Atom *intersectingAtom : intersectingAtoms) {
-		exitDirection += m_pOwnerMO->RotateOffset(intersectingAtom->GetNormal());
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	void AtomGroup::Draw(BITMAP *targetBitmap, const Vector &targetPos, bool useLimbPos, unsigned char color) const {
+		Vector atomPos;
+		Vector normal;
+
+		acquire_bitmap(targetBitmap);
+
+		for (const Atom *atom : m_Atoms) {
+			if (!useLimbPos) {
+				atomPos = (m_OwnerMO->GetPos() + (atom->GetOffset().GetXFlipped(m_OwnerMO->m_HFlipped) * m_OwnerMO->GetRotMatrix()));
+			} else {
+				atomPos = (m_LimbPos + (atom->GetOffset().GetXFlipped(m_OwnerMO->m_HFlipped) * m_OwnerMO->GetRotMatrix()));
+			}
+			// Draw normal first, then draw the Atom position
+			if (!atom->GetNormal().IsZero()) {
+				normal = atom->GetNormal().GetXFlipped(m_OwnerMO->m_HFlipped) * 5;
+				line(targetBitmap, atomPos.GetFloorIntX() - targetPos.GetFloorIntX(), atomPos.GetFloorIntY() - targetPos.GetFloorIntY(), atomPos.GetFloorIntX() - targetPos.GetFloorIntX(), atomPos.GetFloorIntY() - targetPos.GetFloorIntY(), 244);
+			}
+			putpixel(targetBitmap, atomPos.GetFloorIntX() - targetPos.GetFloorIntX(), atomPos.GetFloorIntY() - targetPos.GetFloorIntY(), color);
+		}
+		release_bitmap(targetBitmap);
 	}
 
-    // We don't have a direction to go, so quit
-// TODO: Maybe use previous position to create an exit direction instead then?
-	if (exitDirection.IsZero()) {
-		return false;
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	void AtomGroup::GenerateAtomGroup(MOSRotating *ownerMOSRotating) {
+		BITMAP *refSprite = ownerMOSRotating->GetSpriteFrame();
```

Comment:
This is disgusting and I don't want to read it. Like the travel methods I'd dump a TODO at the top begging for Bjarne to save you.